### PR TITLE
Try/add drag handle to select mode

### PR DESCRIPTION
**Pull request by draganescu**
## Description
Advances #24092

Adds a drag handle to the block selector button when in navigation mode. 

## Screenshots

https://user-images.githubusercontent.com/107534/110501772-2e80a580-8103-11eb-81d5-c9aa3a43c8d7.mp4

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. 
- [ ] My code follows the accessibility standards. 
- [ ] My code has proper inline documentation. 
- [ ] I've included developer documentation if appropriate. 
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR.

### CODE REVIEW

**Comment by jasmussen** [comment]
Whoah. A Sunday PR, and an impressive one at that! This is what I see:

select mode

A few things come to mind, that don't necessarily have to be part of this PR, but are inspired by how good this feels:

1. We should consider adding the block icon before the drag handle in select mode, so that when you lift the block off the page, the black chip that you then see matches in structure.
2. Should the black chip _be_ the "select mode label"? I.e. when in select mode, the block toolbar is the same visual style as the floating chip?

Those are small nitpicks. But one thing I'd love to see, is that the entire _block itself_ is the draggable area, in addition to the drag handle.

selecting

Right now when you click and drag on the block itself, nothing happens. So it seems like it could be the draggable handle instead. The primary tricky thing here is that the 2nd _click_ on the selected block (i.e. click a selected block twice) should still go back to edit mode. So potentially we need to differentiate between ondrag and onclick, and only let the former actually drag.

Nice work! 👌

**Comment by talldan** [comment]
The branch seems to have all the commits from the `try/menus-cleanup` branch. I guess it wasn't based off of master?

**Comment by draganescu** [comment]
>  I guess it wasn't based off of master?
Yes @talldan I was testing the new nav screen design PR and branched from it. I'll fix it but I wanted it to be visible b/c it really feels nice and I was curious if I'm biased :)

**Comment by mtias** [comment]
@jasmussen @pablohoneyhoney would it make sense to use the dark background for the selection state the moment you press "Escape"?

**Comment by jasmussen** [comment]
> would it make sense to use the dark background for the selection state the moment you press "Escape"?

Yes, that would be nice. I can probably push code directly to this branch to that effect?

**Comment by jasmussen** [comment]
I made the handle dark, I think it works well:

dark handle

But now that it's dark, should it be identical to the one you see when you're dragging? I.e. in this case, just show a paragraph icon and a drag handle? (I think yes)

We could also increase the drop shadow _while dragging_, to better ambiguate between resting and lifting. What do you think Andrei?

**Comment by mtias** [comment]
Nice!

> But now that it's dark, should it be identical to the one you see when you're dragging? I.e. in this case, just show a paragraph icon and a drag handle? (I think yes)

Yes, it'd make sense to try. Could also make sense to have [icon] [drag] Label. With label disappearing when you start dragging but the icon and drag handle remaining exactly the same.

**Comment by jasmussen** [comment]
> Yes, it'd make sense to try. Could also make sense to have [icon] [drag] Label. With label disappearing when you start dragging but the icon and drag handle remaining exactly the same.

Yes, that's probably best 👍 👍

**Comment by draganescu** [comment]
Lovely additions @jasmussen

**Comment by jasmussen** [comment]
Made the shadow droppier, and fixed a color issue with the grabby handle active state:

drag shadow

**Comment by draganescu** [comment]
I am thinking of stealing from #23497 to see how that works in this slect mode context.

**Comment by draganescu** [comment]
My opinion is to merge this like it is and follow up with a secondary PR to explore long press drag for the blocks. Should we change the block popover (the handle in navigation mode) to look exactly like the 1st part of the block toolbar? Or is the block name better in this mode than an icon?

**Comment by draganescu** [comment]
@jasmussen I went for (1) [block-type-icon] [drag-handle] Label and also implemented the block selection button to be like the drag chip based on flex, bolted some CSS to make the transition when dragging look normal, and here it is: I updated the screen recording in the PRs description.

What do you think?

File: packages/block-editor/src/components/block-list/style.scss [hunk 1]
```diff
@@ -614,28 +614,58 @@
  */
 
 .block-editor-block-list__block-selection-button {
-	display: block;
+	display: inline-flex;
+	padding: 0 20px;
 	z-index: z-index(".block-editor-block-list__block-selection-button");
+	// Dark block UI appearance.
+	border-radius: $radius-block-ui;
```
Comment (draganescu) [inline packages/block-editor/src/components/block-list/style.scss:621]: The comment seems to be in the wrong place.

File: packages/block-editor/src/components/block-list/style.scss [hunk 2]
```diff
@@ -614,28 +614,58 @@
...
+	height: $block-toolbar-height + 2 * $border-width;
+
+	.block-editor-block-list__block-selection-button__content {
+		margin: auto;
+		display: inline-flex;
+	}
+	.components-button.has-icon.block-selection-button_drag-handle {
+		cursor: grab;
+		padding: 0;
+		height: 24px;
```
Comment (draganescu) [inline packages/block-editor/src/components/block-list/style.scss:635]: I wonder why did I need to bolt these magic numbers here.

File: packages/block-editor/src/components/block-list/style.scss [hunk 3]
```diff
@@ -614,28 +614,58 @@
  */
 
 .block-editor-block-list__block-selection-button {
-	display: block;
+	display: inline-flex;
+	padding: 0 20px;
```
Comment (draganescu) [inline packages/block-editor/src/components/block-list/style.scss:618]: This was kinda too hard to get right, none of the `$grid-unit` options seemed to fit right "visually" with the chip once the drag started.

**Comment by jasmussen** [comment]
This is what I see:

Screenshot 2021-03-09 at 18 31 04

Great. I think I need to push some tweaks to get some of the margin stuff right, it feels slightly off even if it isn't necessarily. I'll massage those pixels tomorrow!

**Comment by jasmussen** [comment]
Alright, pushed a few things to get the metrics right:

Screenshot 2021-03-10 at 08 06 51

Screenshot 2021-03-10 at 08 19 38

drag

I think this is good to land.

**Review by gwwar — APPROVED** [review]
This feels pretty nice and code looks reasonable. Was there anything in particular to focus on here for testing @draganescu ?

For other reviewers 👀 , a quick way of getting into select mode is by pressing <kbd>esc</kbd> while a block is focused.

https://user-images.githubusercontent.com/1270189/110854920-81c13800-826a-11eb-9b3a-c7100e664147.mp4

**Comment by jasmussen** [comment]
🔥

**Comment by mtias** [comment]
Nice!

**Pull request merged**